Security.md

```diff
@@ -0,0 +1,122 @@
+function call with a mismatched signature, exceeding the maximum size of the
+protected call stack, and accessing [out-of-bounds](#Rationale.md#out-of-bounds)
+addresses in linear memory, as well as certain illegal arithmetic operations
+such as division or remainder by zero, and signed division overflow. In the
+future,
+[module-defined trap handlers](FutureFeatures.md#trappingor-non-trapping-strategies)
+will be supported.
+
+### Memory Safety
+Compared to traditional C/C++ programs, these semantics obviate certain classes
+of memory safety bugs in WebAssembly. Buffer overflows, which occur when data
+exceeds the boundaries of an object and accesses adjacent memory regions, cannot
+affect local or global variables stored in index space, they are fixed-size and
+addressed by index. Data stored in linear memory can overwrite adjacent objects,
+since bounds checking is performed at linear memory region granularity and is
+not context-sensitive. However, the presence of control-flow integrity and
+protected shadow call stacks prevents direct code injection attacks. Thus,
+common mitigations such as [data execution prevention][] (DEP) and
+[stack smashing protection][] (SSP) are not useful.
+
+Another common class of memory safety errors involves unsafe pointer usage and
+[undefined behavior](CandC++.md#undefined-behavior). This includes dereferencing
+pointers to unallocated memory (e.g. `NULL`), or freed memory allocations. In
+WebAssembly, the semantics of pointers have been eliminated for function calls
+and variables with fixed static scope, allowing references to invalid indexes in
+any index space to trigger a validation error at load time, or at worst a trap
+at runtime. Accesses to linear memory are bounds-checked at the region level,
+potentially resulting in a trap at runtime. These memory region(s) are isolated
+from the memory space of the runtime, and are set to zero by default unless
+otherwise initialized.
+
+Nevertheless, other classes of bugs are not obviated by the semantics of
+WebAssembly. Although attackers cannot perform direct code injection attacks,
+it is possible to hijack the control flow of a module using code reuse attacks
+(similar to [return-oriented programming][] (ROP)) against indirect calls with
+function-level granularity. Likewise, aside from
+[post-MVP atomic memory primitives](PostMVP.md#threads), the language does not
+provide any execution or scheduling guarantees beyond in-order execution. This
+means that it is possible to write a module that suffers from race conditions,
+such as a [time of check to time of use][] (TOCTOU) vulnerability. Similarly,
+modules can also suffer from [side channel attacks][], such as a timing attack.
+In the future, it may be useful to investigate additional protections, such as
+code diversification or memory randomization (similar to
+[address space layout randomization][] (ASLR)), [bounded pointers][] ("fat"
+pointers), or finer-grained control-flow integrity.
```

Comment:
Code diversification in WebAssembly would seem to be an implementation detail not visible to users of WebAssembly.